### DISCUSSION TITLE
Lifecycle hook between `onBeforeRender` and reply (client navigation)

### DISCUSSION THREAD

**Discussion by chrisvander** [vikejs/vike#1182] [Discussion] [answered]
On the server-side, how can I have an `onBeforeRender` hook that does some transformations to the `pageContext`, and then some default code that runs to do some final changes to `pageContext` before replying to a request on client navigation (e.g. the `index.pageContext.json` routes). Not sure what hook I should use to do this.

**Comment by brillout** [comment]
With the V1 design, you can create [custom hooks using `meta`](https://vike.dev/meta#create-new-configurations).

In the meantime, until you migrated to the V1 design, would calling a function (that performs `some default code that runs to do some final changes`) in each `onBeforeRender()` hook be an option?

**Reply by chrisvander** [reply]
Just finished migrating to the V1 design, actually. No hiccups, nice API. How would I work the custom hook into the lifecycle where I need it?

This would really only save me a line of code in each `onBeforeRender` hook, but I figured I might ask.

So, instead of:
```ts
export default async function onBeforeRender({ trpc }: PageContext) {
  await trpc.site.list.prefetch()
  return { pageContext: { dehydratedQueryState: trpc.dehydrate() } }
}
```

It would be:

```ts
export default async function onBeforeRender({ trpc }: PageContext) {
  await trpc.site.list.prefetch()
}

// before httpResponse is created, but after onBeforeRender hook, and works with `*.pageContext.json` requests
function someLifecycleHook(pageContext: PageContext) {
  return { pageContext: { dehydratedQueryState: pageContext.trpc.dehydrate() } }
}
```

**Reply by brillout** [reply]
You shouldn't mutate `pageContext`, see https://vike.dev/pageContext#lifecycle.

What's wrong with one global `onBeforeRender()` hook (that you define at `pages/+onBeforeRender.js` and is therefore applied to all your pages) that does your trpc integration?

**Reply by chrisvander** [reply]
Sorry, changed the hook to reflect immutability.

Nothing wrong with one global `onBeforeRender()` hook, that might be preferred, except that I need to determine what to prefetch on a page-by-page basis, so I need local `onBeforeRender` hooks. And, I can't specify `onBeforeRender` hooks in multiple places, they won't all run.

**Reply by brillout** [reply, marked as answer]
> except that I need to determine what to prefetch on a page-by-page basis, so I need local `onBeforeRender` hooks.

That's what I thought: that's where a custom hook is convenient &mdash; you'd call that custom hook in your global `onBeforeRender()` hook.

Define a new config `config.myCustomHook()` by using `meta` then:

```ts
// /pages/+onBeforeRender.js

export function onBeforeRender(pageContext) {
   pageContext.config.myCustomHook()
}
```

> And, I can't specify `onBeforeRender` hooks in multiple places, they won't all run.

Yes because the rendering of a page will always only call one `onBeforeRender()` hook (in other words: the global `onBeforeRender()` hooks is overriden by the locale one). That's why we create a custom hook.

**Reply by chrisvander** [reply]
Sounds definitive to me. Ok. So I'll use a global `onBeforeRender` hook to do the dehydration, and a custom hook to do the prefetching. I'll need to remember that any override of `onBeforeRender` will require me to redo the dehydration as well.

The pipe dream would be to somehow figure out what to prefetch by intercepting the `useQuery` calls during SSR in the React component itself. I honestly thought tRPC's NextJS integration would work this way, but was surprised when I saw the prefetch function and dehydration.

But even if I could do that, I don't want to server side render just to figure out what data I need to prefetch and send with the new `pageContext` on client navigation. So, it will remain a pipe dream for the time being.

**Reply by brillout** [reply]
> So I'll use a global `onBeforeRender` hook to do the dehydration, and a custom hook to do the prefetching. 

Yes, it seems to me that it would be the idiomatic way of doing it.

> I'll need to remember that any override of `onBeforeRender` will require me to redo the dehydration as well.

Yes. Alternatively, you never override `onBeforeRender()` and create a bunch of custom hooks (one per need) and orchestrate it all inside your single global `onBeforeRender()` hook. That said, if the logic is fundamentally different, then having more than one `onBeforeRender()` hook can make sense, for example admin pages can have completely different requirements than marketing pages (you'd then have one `onBeforeRender()` hook for admin pages and a second `onBeforeRender()` hook for marketing pages).

> pipe dream

FYI there is a prototype of a deep Telefunc + Vike + React 18 Streaming integration which enables any React component to fetch data. I intend to finish/polish it after `vike@1.0.0` is released.

Also FYI I intend to document the information we're discussing in this thread, which I think is important (but still isn't the top priority at the moment).

**Reply by chrisvander** [reply]
That's what the thread is for!

And yes, I saw Telefunc. Haven't tried it out yet. Heavily invested in tRPC at the moment but I _love_ the API with just a `.telefunc.ts`.
